cpp/Ice/filesystem/Client.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "Filesystem.h"
+
+#include <Ice/Ice.h>
+#include <iostream>
+
+using namespace std;
+using namespace Filesystem;
+
+/// Recursively print the contents of a directory in tree fashion. For files, show the contents of each file.
+/// @param dir The directory to list.
+/// @param depth The current nesting level (for indentation).
+static void
+listRecursive(const DirectoryPrx& dir, size_t depth = 0)
+{
+    const string indent(++depth, '\t');
+
+    // The node proxies returned by list() are never null.
+    NodeSeq contents = dir->list();
+
+    for (const optional<NodePrx>& node : contents)
+    {
+        auto subdir = Ice::checkedCast<DirectoryPrx>(node);
```

Comment:
I wonder if we should just use different categories for the node types, instead of using checked casts.